### ISSUE TITLE
Crash in debug

### ISSUE THREAD

**Issue by yukitoto** [pusher/pusher-websocket-swift#132]
When I try to connect, sometimes crash app because of this library. This issue maybe related to Reachablity. 
![2017-04-17 10 35 48](https://cloud.githubusercontent.com/assets/7534018/25076982/bae375c0-2360-11e7-8663-c3a15090e9d7.png)
![2017-04-17 10 35 58](https://cloud.githubusercontent.com/assets/7534018/25076985/bcf1acec-2360-11e7-8cef-6978cc2d6295.png)

**Comment by hamchapman** [comment]
@yukitoto that will almost certainly be happening because you're not retaining your instance of `Pusher` anywhere and so `self` (your instance of `Pusher`) is `nil` when the `Reachability` block at the top of `PusherConnection.swift` is run. It's very similar to this: https://github.com/pusher/pusher-websocket-swift/issues/109 issue.

The short summary is that you should be able to just keep a reference to your instance of `Pusher` and you won't get the error anymore.

I'll be looking at either making the behaviour for this part of the library change or document things to make the expected usage clearer.

**Comment by yukitoto** [comment]
@hamchapman  Thank you for your reply. However I already retain the instance of `Pusher`.
This is my code. This crash occur when call the function named `initializePusherIfNeeded`.
Is there any mistake?

```
struct PusherManager {
    static fileprivate var pusher: Pusher?
    
    
    // MARK: - Connection -
    
    fileprivate static func initializePusherIfNeeded() {
        if pusher != nil { return }
        
        let options = PusherClientOptions(
            autoReconnect: true,
            host: .cluster("ap1"),
            encrypted: true)
        
        pusher = Pusher(
            key: "fffffffffffff",
            options: options)
    }
}
```

**Comment by yukitoto** [comment]
@hamchapman We have some crashes caused by this issue. Is there any workaround or something?

**Comment by hamchapman** [comment]
The error suggests that something isn't being retained, presumably the `Reachability` object. I think this would only be the case if the `Pusher` instance isn't being retained properly either.

How are you storing a reference to your `PusherManager` object?

**Comment by yukitoto** [comment]
@hamchapman  `static fileprivate var pusher: Pusher?` is not a instance property. So if I do not have a reference to my `PusherManager`, I think that it  is not released,.

**Comment by hamchapman** [comment]
Good point, sorry I missed that.

I can't seem to recreate the crash locally - do you have a reproducible method for generating the crash?

Thanks!

**Comment by yukitoto** [comment]
I don't know why, but when I changed the code like below, this issue is fixed.
```
public struct PusherManager {
    
    // MARK - Properties -

    static fileprivate let pusher: Pusher = {
        let options = PusherClientOptions(
            autoReconnect: true,
            host: .cluster("ap1"),
            encrypted: true)
        
        return Pusher(
            key: InfoPlistUtil.get(type: .pusherAccessKey)!,
            options: options)
    }()
```

**Comment by hamchapman** [comment]
Please can you try updating to 4.2.1 (or 5.0.0beta1 if you're using Swift 4) to see if this fixes the problem you had originally?

If not, I'll consider this closed as I think it was the same issue.

Thanks!

**Comment by hamchapman** [comment]
Closing this as I think the issue is fixed since 4.2.1.
